hudi-spark/src/main/scala/org/apache/hudi/HudiMergeOnReadRDD.scala

```diff
@@ -0,0 +1,195 @@
+  }
+
+  private def unMergeFileIterator(split: HudiMergeOnReadFileSplit,
+                                  readFileFunction: PartitionedFile => Iterator[Any]): Iterator[InternalRow] =
+    new Iterator[InternalRow] {
+      private val dataFileIterator = read(split.dataFile, readFileFunction)
+      private val logSchema = getLogAvroSchema(split)
+      private val sparkTypes = SchemaConverters.toSqlType(logSchema).dataType.asInstanceOf[StructType]
+      private val converter = new AvroDeserializer(logSchema, sparkTypes)
+      private val hudiLogRecords = scanLog(split, logSchema).getRecords
+      private var hudiLogRecordsIterator = hudiLogRecords.keySet().iterator().asScala
+
+      override def hasNext: Boolean = {
+        dataFileIterator.hasNext || hudiLogRecordsIterator.hasNext
+      }
+
+      override def next(): InternalRow = {
+        if (dataFileIterator.hasNext) {
+          dataFileIterator.next()
+        } else {
+          val curAvrokey = hudiLogRecordsIterator.next()
+          val curAvroRecord = hudiLogRecords.get(curAvrokey).getData.getInsertValue(logSchema).get()
+          converter.deserialize(curAvroRecord).asInstanceOf[InternalRow]
+        }
+      }
+    }
+
+  private def mergeFileIterator(split: HudiMergeOnReadFileSplit,
+                                readFileFunction: PartitionedFile => Iterator[Any]): Iterator[InternalRow] =
+    new Iterator[InternalRow] {
+      private val dataFileIterator = read(split.dataFile, readFileFunction)
+      private val logSchema = getLogAvroSchema(split)
+      private val sparkTypes = SchemaConverters.toSqlType(logSchema).dataType.asInstanceOf[StructType]
+      private val converter = new AvroDeserializer(logSchema, sparkTypes)
+      private val hudiLogRecords = scanLog(split, logSchema).getRecords
+      private var parquetFinished = false
+      private var logRecordToRead = hudiLogRecords.keySet()
+      private var hudiLogRecordsIterator: Iterator[String] = _
+
+      override def hasNext: Boolean = {
+        if (dataFileIterator.hasNext) {
+          true
+        } else {
+          if (!parquetFinished) {
+            parquetFinished = true
+            hudiLogRecordsIterator = logRecordToRead.iterator().asScala
+          }
+          hudiLogRecordsIterator.hasNext
+        }
+      }
+
+      override def next(): InternalRow = {
+        if (dataFileIterator.hasNext) {
+          val curRow = dataFileIterator.next()
+          val curKey = curRow.getString(HOODIE_RECORD_KEY_COL_POS)
+          if (hudiLogRecords.keySet().contains(curKey)) {
+            logRecordToRead.remove(curKey)
+            mergeRowWithLog(curRow)
+          } else {
+            curRow
+          }
+        } else {
+          val curKey = hudiLogRecordsIterator.next()
+          getAvroRecord(curKey)
+        }
+      }
+
+      private def getAvroRecord(curKey: String): InternalRow = {
+        val curAvroRecord = hudiLogRecords.get(curKey).getData.getInsertValue(logSchema).get()
```

Comment:
this implicitly assumes `OvewriteWithLatestPayload` ? can we just convert the parquet row as well to Avro and then perform the merge actually calling the right API. `HoodieRecordPayload#combineAndGetUpdateValue()` ?  This is a correctness issue we need to resolve in the PR.. 
ideally, adding a test case as well to go along with this would be good

Reply:
Will do.